src/transport/TCP.cpp

```diff
@@ -0,0 +1,585 @@
+            // This is necessary in the case where the message size exceeds the TCP window size to ensure
+            // the peer has enough window to send us the entire message.
+            uint16_t neededLen = messageSize - buffer->DataLength();
+            err                = endPoint->AckReceive(neededLen);
+            SuccessOrExit(err);
+        }
+
+        // Buffer is incomplete and we cannot get more data
+        break;
+    }
+
+exit:
+    if (buffer != nullptr)
+    {
+        // Incomplete processing will be retried
+        endPoint->PutBackReceivedData(buffer);
+    }
+
+    return err;
+}
+
+void TCPBase::OnTcpReceive(Inet::TCPEndPoint * endPoint, System::PacketBuffer * buffer)
+{
+    IPAddress ipAddress;
+    uint16_t port;
+
+    endPoint->GetPeerInfo(&ipAddress, &port);
+    PeerAddress peerAddress = PeerAddress::TCP(ipAddress, port);
+
+    TCPBase * tcp  = reinterpret_cast<TCPBase *>(endPoint->AppState);
+    CHIP_ERROR err = tcp->ProcessReceivedBuffer(endPoint, peerAddress, buffer);
+
+    if (err != CHIP_NO_ERROR)
+    {
+        // Connection could need to be closed at this point
+        ChipLogError(Inet, "Failed to receive TCP message: %s", ErrorStr(err));
+    }
+}
+
+void TCPBase::OnConnectionComplete(Inet::TCPEndPoint * endPoint, INET_ERROR inetErr)
+{
+    CHIP_ERROR err          = CHIP_NO_ERROR;
+    bool foundPendingPacket = false;
+    TCPBase * tcp           = reinterpret_cast<TCPBase *>(endPoint->AppState);
+    IPAddress ipAddress;
+    uint16_t port;
+
+    endPoint->GetPeerInfo(&ipAddress, &port);
+    PeerAddress addr = PeerAddress::TCP(ipAddress, port);
+
+    // Send any pending packets
+    for (size_t i = 0; i < tcp->mPendingPacketsSize; i++)
+    {
+        if ((tcp->mPendingPackets[i].peerAddress != addr) || (tcp->mPendingPackets[i].packetBuffer == nullptr))
+        {
+            continue;
+        }
+        foundPendingPacket = true;
+
+        if ((inetErr == CHIP_NO_ERROR) && (err == CHIP_NO_ERROR))
+        {
+            err = endPoint->Send(tcp->mPendingPackets[i].packetBuffer);
+        }
+
+        tcp->mPendingPackets[i].packetBuffer = nullptr;
+        tcp->mPendingPackets[i].peerAddress  = PeerAddress::Uninitialized();
+    }
+
+    if (err == CHIP_NO_ERROR)
+    {
+        err = inetErr;
+    }
+
+    if (!foundPendingPacket && (err == CHIP_NO_ERROR))
+    {
+        // Force a close: new connections are only expected when a
+        // new buffer is being sent.
+        ChipLogError(Inet, "Connection accepted without pending buffers");
+        err = CHIP_ERROR_CONNECTION_CLOSED_UNEXPECTEDLY;
+    }
+
+    // cleanup packets or mark as free
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Inet, "Connection complete encountered an error: %s", ErrorStr(err));
+        endPoint->Free();
+    }
+    else
+    {
+        for (size_t i = 0; i < tcp->mActiveConnectionsSize; i++)
+        {
+            if (tcp->mActiveConnections[i] == nullptr)
```

Comment:
Our logic should ensure this would never happen, however I added a check with log in case we do have logic errors. The count tracking is a bit on the complex side, especially in unexpected error conditions (happy path is unit tested already)